### DISCUSSION TITLE
Exception java.lang.NoSuchMethodError: uk.ac.manchester.tornado.api.exceptions.Debug.tprintf(Ljava/lang/String;[Ljava/lang/Object;)

### DISCUSSION THREAD

**Discussion by andrii0lomakin** [beehive-lab/TornadoVM#475] [Q&A] [answered]
Hi.

I am working on implementation of the LLM inference engine using TornadoVM, and of course, I had to create a test configuration that runs tests inside the Gradle.

I have generated JVM parameters according to the source code of the `tornado` Python script, which is listed in this gist: https://gist.github.com/andrii0lomakin/d4e8893e46d4ee1a66820cba25e5358f.

And though the `showDevices` task works perfectly well. 
During the creation of the task graph I have got the following:

```
java.lang.ExceptionInInitializerError: Exception java.lang.NoSuchMethodError: uk.ac.manchester.tornado.api.exceptions.Debug.tprintf(Ljava/lang/String;[Ljava/lang/Object;)
```

Which seems logical because this method does absent in provided class.
But it also means that I miss something could you help me with this problem ?

**Comment by jjfumero** [comment]
Hi @andrii0lomakin . I have little experience with Gradle, but just a few things that can help:

1. Are you using all Java flags printed by the `tornado` script? 

For example, for my configuration using the OpenCL backend, I got all these flags: 

```bash
$ tornado --printJavaFlags 

/home/juan/repos/tornadovm/etc/dependencies/TornadoVM-graalvm-jdk-21/graalvm-community-openjdk-21.0.1+12.1//bin/java -server -XX:-UseCompressedOops -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -XX:-UseCompressedClassPointers --enable-preview -Djava.library.path=/home/juan/repos/tornadovm/bin/sdk/lib  --module-path .:/home/juan/repos/tornadovm/bin/sdk/share/java/tornado -Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph -Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime -Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado -Dtornado.load.annotation.implementation=uk.ac.manchester.tornado.annotation.ASMClassVisitor -Dtornado.load.annotation.parallel=uk.ac.manchester.tornado.api.annotations.Parallel  -XX:+UseParallelGC  @/home/juan/repos/tornadovm/bin/sdk/etc/exportLists/common-exports @/home/juan/repos/tornadovm/bin/sdk/etc/exportLists/opencl-exports --add-modules ALL-SYSTEM,tornado.runtime,tornado.annotation,tornado.drivers.common,tornado.drivers.opencl
```

2. In addition, use the same `JAVA_HOME` as tornado. After the TornadoVM installation, the installer generates a conf file called (`setvars.sh`). You can copy your `PATH` and `JAVA_HOME` from this source file.  

Let us know if this helps.

**Reply by andrii0lomakin** [reply]
Defaulting to the TornadoVM java home helped. I use Temurine JDK 21 by default. Could you explain why?  I thought that TornadoVM was a JIT extension + JNI libraries to call native drivers. What is limiting it to use only this type of JDK? It limits its usage in such cases. Probably, I merely need to copy config files from TornadoVM JDK home, which I am not aware of ?

**Reply by andrii0lomakin** [reply]
Well, my bad, it did not help. Tests failed on another error I accidentally introduced. I will investigate it further. I appreciate your help.

**Reply by jjfumero** [reply]
So, you can run TornadoVM with any compatible JDK. Indeed, TornadoVM uses Graal as a JAR library and extends it with new compilation phases, backends and new nodes in the IR.  There is no need to use the same as used to configure TornadoVM. But, the only thing is that the JVMCI [1] version might change across versions (thus, the Graal JIT Compiler interface). So just to be sure, you can use the same. 

[1] https://openjdk.org/jeps/243

**Reply by andrii0lomakin** [reply]
Yes, I use the same. Running the same code outside of the JUnit executor works perfectly fine. So it looks like a problem inside of it.

**Reply by andrii0lomakin** [reply]
@jjfumero, my best guess is that TornadoVM uses an incorrect classloader to add uk.ac.manchester.tornado.drivers.opencl.graal.nodes.TPrintfNode if class loaders are overridden by the application. Could you suggest to me how I can debug this case from my application? Initially, I thought to convert everything to maven. Still, again if you use TornadoVM as a JIT extension for a bigger application that uses Gradle for testing, that is a no-go. So I want to resolve this issue on my current environment.

**Reply by jjfumero** [reply]
Are you using Java modules in your application? Perhaps TornadoVM is not exporting some modules correctly.  If you can provide a test-case, we can take a look.

**Reply by andrii0lomakin** [reply]
No, I definitely do not use modules. Also, if I want to contribute, I need to know how to debug library :-) So again, if you provide me with a short guideline, I would be grateful to fix this issue. Considering our plans and the state of the library (it requires some polishing), it will not be a single-time contribution.

**Reply by andrii0lomakin** [reply]
As for the class loader problem I am quite confident in it, the only difference between standalone run and run with gradle test runner is the fact that it wraps several classloader, all jvm arguments completely the same except of `-ea`

**Reply by jjfumero** [reply]
I would say that the first thing is to build a minimal test case with minimal dependencies that breaks, so it will be easier to analyze. 

For the contributions you are very welcome. It depends on the part you want to contribute: codegen,  optimizations, etc. We use different debugging tools for each component.

**Reply by andrii0lomakin** [reply]
@jjfumero I found the reason.
In the case of disabled assertions this line https://github.com/oracle/graal/blob/release/graal-vm/23.1/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/nodes/graphbuilderconf/InvocationPlugins.java#L805 is skipped and code executed correctly but in case of enabled assertions GraalVM throws exception that provided plugin does not satisfy requirements. So, I have a couple of questions:

1. Do you run your tests with -ea enabled? It is strange that they passed on your side, so I had to ask.
3. In code, you provide plugins for nonexisting methods , such as tprintf or Math.pow(float, float). Do you really need them?  I mean, code is never generated for them if you do not call them in Java bytecode, right? So why are they needed?

**Reply by andrii0lomakin** [reply]
Namely, this part never finds declared methods:
```
      Method match = null;
        for (Method m : methods) {
            if (plugin.isSameType(m)) {
                if (match == null) {
                    match = m;
                } else if (match.getReturnType().isAssignableFrom(m.getReturnType())) {
                    // `m` has a more specific return type - choose it
                    // (`match` is most likely a bridge method)
                    match = m;
                } else {
                    if (!m.getReturnType().isAssignableFrom(match.getReturnType())) {
                        throw new NoSuchMethodError(String.format(
                                        "Found 2 methods with same name and parameter types but unrelated return types:%n %s%n %s", match, m));
                    }
                }
            }
        }

**Reply by andrii0lomakin** [reply]
I made a full text scan and did not find that `tprintf` is called in the text except of GPU "byte code" but AFAI GPU byte code generated by GraalVM but not converted it into the  real executable code and that is responsibility of CUDA and co.

**Reply by jjfumero** [reply]
Yes. We run our uni-tests with `-ea` enabled. 
Regarding the Graal compiler plugins, we need this to insert new IR nodes into Graal IR during the Graal Graph Build Process. In this way, we start mapping the semantics of some operations down to the GPU. This includes a wide number of operations and methods (e.g., access to Panama regions, vector operations, vector types, Math operations, etc). Plugins are common for all TornadoVM backends, and then, when you specialise, each backend lowers IR nodes to backend-specific operations (e.g., CUDA PTX, or SPIR-V, etc). 

In the case of Math operations with `float` for example, we map this to `float` parameters, which is legal in OpenCL, PTX and SPIR-V. Otherwise, Graal and TornadoVM will do a typecast from `double` o `float`, operate and then do the typecast again. However, some GPUs (e.g,, Intel ARC, Intel HD Graphics, and some NVIDIA GPUs such as Jetson Nano) do not support FP64. Thus, we provide an interface to be plugin directly during the Graph Builder to replace the `double` signature with `float` signature. In this case, typecast is gone and we can operate on GPUs without FP64 support.

**Reply by andrii0lomakin** [reply]
I understand all of this. My question is different.
If you provide a plugin for Math.pow(float, float) and this code does not exist (the method is absent from the user's perspective), the user will never write it. How will the plugin be used and called?

**Reply by andrii0lomakin** [reply]
>Yes. We run our uni-tests with -ea enabled.

Got it. Is it enough for me to call a test task in Maven to run those tests ? (asking to save time for configuration of environment)

**Reply by jjfumero** [reply]
> I understand all of this. My question is different. If you provide a plugin for Math.pow(float, float) and this code does not exist (the method is absent from the user's perspective), the user will never write it. How will the plugin be used and called?

The plugin should be for `TornadoMath.pow`, otherwise we have a bug.Let me check.

**Reply by jjfumero** [reply]
> > Yes. We run our uni-tests with -ea enabled.
> 
> Got it. Is it enough for me to call a test task in Maven to run those tests ? (asking to save time for configuration of environment)

Yes, that should be fine. Thanks

**Reply by andrii0lomakin** [reply]
@jjfumero I hardcoded " -ea " flag into the `tornado` script and then startet "tornado-test" script.
Unfortonatelly after that change all tests start to fail.
Here is the output https://gist.github.com/andrii0lomakin/7e4ef73820008c8a3e671a966a21a280 
Here is output of `tornado --printJavaFlags` - https://gist.github.com/andrii0lomakin/c9bb7739cf1ac7262a3cc9b48f554175
Here is changed `tornado` script https://gist.github.com/andrii0lomakin/dcce3ca9b2750d8b50000616fd368988
Here is output of `tornado --devices` with `-ea` enabled https://gist.github.com/andrii0lomakin/86d668970da9d0d6b6060272914b07d1
and with `-ea` disabled https://gist.github.com/andrii0lomakin/d0d50bf57d7fe6ecc8e09aa8781fd0a7

Are you sure that your run tests with `-ea` enabled taken all those facts into account?

**Reply by andrii0lomakin** [reply]
My preliminar impression at the moment is that you have some plugins registered incorrectly. Will be happy to be disproved.

**Reply by andrii0lomakin** [reply]
If you agree, I already started to work on the modified version because that is a show stopper for me, and I will be glad to provide PR.

**Reply by andrii0lomakin** [reply]
> > I understand all of this. My question is different. If you provide a plugin for Math.pow(float, float) and this code does not exist (the method is absent from the user's perspective), the user will never write it. How will the plugin be used and called?
> 
> The plugin should be for `TornadoMath.pow`, otherwise we have a bug.Let me check.

I do not remember exact place (this concrete one related to PTX driver), there are several ones, probably better for you do not waste time, information is not very concrete I will provide PR soon and you will be able to check code.

**Reply by jjfumero** [reply, marked as answer]
Can you run with these flags?


```
# We do not satisfy the Graal compiler assertions because we only support a subset of the Java specification.
# This allows us to have the GraalIR in states which normally would be illegal.
__GRAAL_ENABLE_ASSERTIONS__ = " -ea -da:org.graalvm.compiler... "
```

`tornado -ea` enables assertions and disables the ones for the Graal Compiler. 

So you can use:

```
--jvm="-ea -da:org.graalvm.compiler..."
```

**Reply by andrii0lomakin** [reply]
I see, I suppose it will work in such case (I will check). In such case you also skip valid assertions as the ones related to creation of plugins to the methods that existed before but disappeared not valid anymore. Like ones that I listed above. Let me try anyway.

**Reply by andrii0lomakin** [reply]
Yes, it works but as I have noticed it will skip valid cases too. I will try to gather such cases and create separate issue they are not critical obviously. I have another issue,  TornadoVM does not work with Kotlin intrisincs (does not take into account that for them trusted stamp can be null during parsing of byte code) . That is small fix I will provide it soon.

**Reply by andrii0lomakin** [reply]
IMHO would be good to create small a guideline how to run test if TornadoVM is part of bigger application.  If I have time I will help with that.

**Reply by jjfumero** [reply]
Thanks. That will be great.

**Reply by andrii0lomakin** [reply]
Conclusion:
 The problem was caused by  the mix:
 
 1. Absence of -da in tests.
 2. Fact that Kotlin adds a. intrisincs b. Array.rangeCopy in lamba so it is strongly recommended do not pass Koltin lambadas.
 
Action points:
1. If I have time, I will provide a list of plugins that do not do anything useful or methods that are expected in TornadoMath but are absent. 
2. Provide PR to handle Kotlin intrinsics correctly. It will only make processing more stable, though it does not solve the problem with Array.rangeCopy

Updates:
I have checked, and for the moment, the removal of disabling of compressed pointers does not bring any harm.

**Reply by andrii0lomakin** [reply]
As for the last, I will create a separate discussion.

**Comment by jjfumero** [comment]
I forgot to mention that the flags depend on the backend you have installed. The ones I showed are specific for the OpenCL backend and my configuration. If you compiled with the PTX or SPIR-V, you will see a couple of different flags. The way to query is via 

```bash
$ tornado --printJavaFlags 
```

**Reply by andrii0lomakin** [reply]
@jjfumero, thank you for your help. I am aware of this. I converted the `tornado` Python script into Gradle script line-by-line.

**Comment by jjfumero** [comment]
For reference, @mikepapadim , core member of the TornadoVM team, extended LLM Llama2.java with TornadoVM: 

https://github.com/mikepapadim/llama2.tornadovm.java  

This project uses Maven, but it can serve as reference.

**Reply by andrii0lomakin** [reply]
I appreciate the reference. It is useful. But in general, I find the JLama design, not GPU performance-friendly. I hope to achieve a better speed-up. 1.15x  speedup is not motivating users to move to quite a specific library implementation.

**Reply by jjfumero** [reply]
Indeed. So 15% improvements is just because 1 of the compute kernels was rewritten to use TornadoVM (the matrix-vector kernel). What Michalis found is that the data transfers back and forth (host <-> device) are now the bottleneck, and we are working on a new version to improve TornadoVM with automatic shared memory.

**Reply by andrii0lomakin** [reply]
Yeah, I expected that host-device transfer would be a bottleneck. I am working on a design that avoids that from scratch. I also work on other optimizations like Flashed Attention 2.0 and so on. That is why I raised the question of support of tensor cores in TornadoVM.

**Reply by andrii0lomakin** [reply]
Hopefully, we will provide this contribution later after the initial inference implementation is ready.

**Reply by jjfumero** [reply]
> Yeah, I expected that host-device transfer would be a bottleneck. I am working on a design that avoids that from scratch. I also work on other optimizations like Flashed Attention 2.0 and so on. That is why I raised the question of support of tensor cores in TornadoVM.

That would be awesome. Keep us posted about your progress.

**Comment by andrii0lomakin** [comment]
BTW you disable compressed pointers. Is this option really needed? Imagine an application that uses TornadoVM as an embedded inference engine; would it not cause an explosion in heap consumption?

**Reply by andrii0lomakin** [reply]
I mean native memory consumption of course.

**Reply by jjfumero** [reply]
Good catch. I think we do not need this in TornadoVM anymore. So you can disable it.

**Reply by andrii0lomakin** [reply]
I will try it once I have more or less a working implementation and will provide my first meaningful PR :-)
